ddl/ddl_api.go

```diff
@@ -5128,3 +5136,116 @@ func (d *ddl) AlterIndexVisibility(ctx sessionctx.Context, ident ast.Ident, inde
 	err = d.callHookOnChanged(err)
 	return errors.Trace(err)
 }
+
+func validatePlacementSpecs(specs []*ast.PlacementSpec) ([]*placement.Rule, error) {
+	rules := make([]*placement.Rule, 0, len(specs))
+	for k, spec := range specs {
+		rule := &placement.Rule{
+			ID:       placementRuleDefaultID,
+			Count:    int(spec.Replicas),
+			Override: true,
+		}
+
+		switch spec.Tp {
+		case ast.PlacementAdd:
+		default:
+			return rules, errors.Errorf("invalid placement spec[%d], unknown action type: %d", k, spec.Tp)
+		}
+
+		switch spec.Role {
+		case ast.PlacementRoleFollower:
+			rule.Role = placement.Follower
+		case ast.PlacementRoleLeader:
+			rule.Role = placement.Leader
+		case ast.PlacementRoleLearner:
+			rule.Role = placement.Learner
+		case ast.PlacementRoleVoter:
+			rule.Role = placement.Voter
+		default:
+			return rules, errors.Errorf("invalid placement spec[%d], unknown role: %d", k, spec.Role)
+		}
+
+		for _, label := range strings.Split(spec.Constraints, ",") {
```

Comment:
On second thought, `CONSTRAINTS` can be ignored. For example, users just want to specify that 5 replicas are needed, regardless of the places where the replicas will be scheduled to.

Reply:
https://github.com/pingcap/tidb/pull/18611/commits/b3cf7f9259435a224d11013966008aac7e5bd716